### ISSUE TITLE
FreeBSD target: service: name=postgresql state=* hangs the whole playbook

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#30001]
##### ISSUE TYPE

bug report

##### COMPONENT NAME

service module

##### ANSIBLE VERSION

ansible 1.9.4

##### CONFIGURATION

retry_files_enabled = False

##### OS / ENVIRONMENT

Running from Mac OS X 10.10.5 python 2.7.10
to FreeBSD 10.1-RELEASE-p16 0 Tue Jul 28 120419 UTC 2015
with PostgreSQL 9.4.5

##### SUMMARY

Attempting to use the 'service' module to control PostgreSQL hangs every time. I strongly suspect that this is the same issue as https://github.com/ansible/ansible/issues/5923

##### STEPS TO REPRODUCE

```
1) Install postgres94-client and postgres94-server packages on FreeBSD 10.1
2) set up inventory so that the dbserver is in group fll2-dev-dbservers in inventory file development.ini
[fll2-dev-dbservers]
fll2gdbs01dev

3) ansible -vvvv -i development.ini fll2-dev-dbservers -m service -a "name=postgresql state=stopped"
```

##### EXPECTED RESULTS

pg_ctl is called to stop the server, returns when finished, ansible reports success/failure and server state

##### ACTUAL RESULTS

```
<fll2gdbs01dev> ESTABLISH CONNECTION FOR USER: rmartin
<fll2gdbs01dev> REMOTE_MODULE service name=postgresql state=restarted
<fll2gdbs01dev> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/rmartin/.ansible/cp/ansible-ssh-%h-%p-%r" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 fll2gdbs01dev /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931 && echo $HOME/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931'
<fll2gdbs01dev> PUT /var/folders/j_/gh1x7cn138b04zb7w6q41ffw59y07m/T/tmpHfd5G9 TO /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/service
<fll2gdbs01dev> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/rmartin/.ansible/cp/ansible-ssh-%h-%p-%r" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 fll2gdbs01dev /bin/sh -c 'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/local/bin/python2.7 /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/service; rm -rf /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/ >/dev/null 2>&1'
```

<hangs here>

ps auxww excerpt on the server during the hang:

```
rmartin 1258   0.0  0.1 17068  2484  0  Is+   3:14PM  0:00.01 /bin/sh -c LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/local/bin/python2.7 /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/service; rm -rf /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/ >/dev/null 2>&1
rmartin 1259   0.0  0.3 81724 13236  0  S+    3:14PM  0:00.16 /usr/local/bin/python2.7 /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/service
rmartin 1261   0.0  0.1 17068  2852  0  I+    3:14PM  0:00.03 /bin/sh /usr/local/etc/rc.d/postgresql onestatus
root    1268   0.0  0.1 49768  2652  0  I<+   3:14PM  0:00.00 su -l pgsql -c exec /usr/local/bin/pg_ctl -D /usr/local/pgsql/data -l /dev/null -w -s -m fast status
```

I have tried both with and without the fix suggested in the old ticket #5923:

```
postgresql_flags="-l /dev/null -w -s -m fast"
```

It has not made a difference.

Note that copying the ansible command 

```
LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/local/bin/python2.7 /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/service; rm -rf /home/rmartin/.ansible/tmp/ansible-tmp-1447359262.53-79308781944931/ >/dev/null 2>&1
```

from the hung SSH command above, and pasting it into a bash prompt running on that server over interactive SSH works fine.

Any ideas why this is misbehaving, or how to work around it?

Copied from original issue: ansible/ansible-modules-core#2459

**Comment by ansibot** [comment]
From @bcoca on 2015-11-12T20:22:44Z

hm, thought this had been fixed, but it was closed due to the module split and never reopened. 

So last i checked the issue was that EOF was not being sent and Ansible hung waiting for it, need to check why the timeout does not kick in. Also might need a 'badly behaved' services list that does this and try to deal with it more cleanly (timeout will result in error though service is started).

**Comment by ansibot** [comment]
From @rmartin7c on 2015-11-12T20:22:44Z

```
ssh -t user@dbserver "su root -c 'service postgresql start|stop'" 
```

works fine, so the postgresql service is well-behaved enough for ssh. What else is the ansible service wrapper that gets copied to the remote doing, that it would be more sensitive?

**Comment by ansibot** [comment]
From @bcoca on 2015-11-12T20:22:44Z

it does a wait on stdout/stderr, if it does not receive an EOF (which I believe was the issue in the previous ticket) it SHOULD be stuck until the timeout hits. 

In 2.0 we've fixed several timeout issues that might show this as an error instead of getting stuck, going into Ansiblfest 'mode' now so I won't be able to test this until next week.

**Comment by ansibot** [comment]
From @electrofelix on 2015-11-12T20:22:44Z

Looking at the service module code I've noticed a few things that could be contributing here as well as causing other issues.
1. The code at https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L224-L226 suffers from a similar programming bug as https://github.com/ansible/ansible/issues/15436 in that it has a race in how it checks for data from the subprocess and when it checks for it having exited and not checking if any data was sent in between.
2. https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L239-L241 will shutdown the forked process in a manner that does not flush stdio, and doesn't explicitly flush the write on the pipe, it probably should call `os.fsync(pipe[1])` before the `os.close(pipe[1])`.
3. https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L193-L203 looks to duplicate and close some file descriptors, however it should probably look to close all file descriptors that are not `pipe[1]` using something like http://code.activestate.com/recipes/278731-creating-a-daemon-the-python-way/ and then perform the necessary redirection to /dev/null.
4. https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L218 launches the process and uses a preexec_fn to close the pipe file descriptor in the subprocess, it should probably just use `close_fds=True` instead.

All these combined seem like they would provide opportunity for incorrectly and unexpected behaviour. Looking at the ansible 2.x code, it appears to be identical, so the same problem exists there.

As an aside, I'm wondering why this code needs to daemonize at all? I'm guessing some services are not great at closing off their input/output file descriptors leading to some interesting behaviour if you just try to use subprocess to trigger a service restart. However it would seem like the use of `close_fds=True` would remove all of that and ensure that the file descriptors got closed off by the subprocess module (which has been around since 2.4), and seems to have been specifically put in place to handle such situations. Possibly the fix for 2.x is to switch to using that and drop the need to double fork, and for 1.9 to patch the existing behaviour to get the pipes working correctly?

Since this is impacting locally I'm going to see what happens when I change around the 4 pieces of code mentioned above and see if it's possible to fix this (have 1-2 occurrences per day triggered by CI locally).

**Comment by ansibot** [comment]
From @electrofelix on 2015-11-12T20:22:44Z

I forgot to mention also:
1. https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L246-L255 should follow the similar pattern of the code at https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L223-L236 (with fixes to avoid the race) of checking whether the process is running or not in order to help exit the loop and also avoid the scenario that enough output comes from the forked child process that might cause it to fill the pipe and block on the os.write() thus causing the parent to wait forever on `os.waitpid(pid, 0)`. Can use `os.kill(pid, 0)` to check if a pid refers to a running process.

These blocks of code seem similar enough, along with what is used by the ssh connection plugin that they could be placed into a function that is used everywhere this pattern of communication is needed. A function doing something of the following would probably help

```
def get_output(fds, process_running_callback, block_size=4096)
    """General function to handle reading from process file descriptors

    fds is a list of file descriptors to check for data to read
    process_running_callback is the lambda function that
        returns False if the process is finished and True otherwise

   returns a list containing the output read from each fds passed in,
       ordered same as the input
    """
```

**Comment by ansibot** [comment]
From @geerlingguy on 2015-11-12T20:22:44Z

I've been debugging a similar issue with Ansible running inside Docker containers in Docker 1.12 (haven't tested other recent versions, but running inside Docker 1.9 doesn't exhibit the same issue) with sysvinit instead of systemd.

It's a pernicious bug, but it seems to affect some of my roles and not others. Most recently, I noticed it with my Memcached role—see failed build: https://travis-ci.org/geerlingguy/ansible-role-memcached/builds/159737521

The playbook run succeeds, but for some reason `ansible-playbook` never exits on the Ubuntu 14.04 and 12.04 containers (it does on 16.04 and CentOS 7, which use systemd/different init system). I've reproduced the issue locally, and if you have Docker for Mac, you can do the same:
1. `git clone https://github.com/geerlingguy/ansible-role-memcached.git`
2. `cd ansible-role-memcached`
3. `docker pull geerlingguy/docker-ubuntu1404-ansible:latest`
4. `docker run --detach --volume=\`pwd`:/etc/ansible/roles/role_under_test:ro geerlingguy/docker-ubuntu1404-ansible:latest /sbin/init`
5. `docker exec --tty [CONTAINER_ID_HERE] env TERM=xterm ansible-playbook /etc/ansible/roles/role_under_test/tests/test.yml --syntax` (Do a syntax check - this succeeds).
6. `docker exec --tty [CONTAINER_ID_HERE] env TERM=xterm ansible-playbook /etc/ansible/roles/role_under_test/tests/test.yml` (This runs the playbook to the end, but then it hangs after the play summary prints).

If I set the role to not use the `service` module at all, then docker exec returns to the prompt correctly. If `service` is used, the playbook run stalls, you have to press `CTRL^C` to exit docker exec, then you have to restart Docker and finally run `docker rm -fv [CONTAINER_ID_HERE]` to delete the broken container.

Note that if I remove the `--tty` option, then the play succeeds. Note also that memcached _doesn't_ successfully start (if I manually run `status` I get `* memcached is not running`...), so I'm 99% sure it's memcached's fault, but taking over the TTY is a bit annoying, and it would be nice if Ansible could finish it's playbook run and not hang the terminal.

(Also, if I check Memcached's own logs in the container at `/var/log/memcached.log`, I'm getting `failed to listen on TCP port 11211: Address already in use`, and that seems to be what's preventing it from starting—must be two running processes clashing).

**Comment by ansibot** [comment]
From @geerlingguy on 2015-11-12T20:22:44Z

Note that if I downgrade to Docker 1.9.1, even with the `--tty` option, the play succeeds and exits correctly: https://travis-ci.org/geerlingguy/ansible-role-memcached/jobs/159755194

However, in that build, in the Ubuntu 14.04/12.04 runs, you can see that the `service` task is the one that breaks idempotence (https://travis-ci.org/geerlingguy/ansible-role-memcached/jobs/159755194#L512) — since Memcached isn't successfully/correctly starting in the first place.

(Also, after more debugging, if I restart the container (`docker stop` and `docker start`) then run the playbook, everything works correctly. It seems that on the initial `service` module call something with the init script messes up the start of memcached to the point where Netstat shows something's listening on 11211 but doesn't know what (this is what makes memcached's init script fail to exit correctly):

```
$ docker exec [CONTAINER_ID_HERE] env TERM=xterm netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:11211         0.0.0.0:*               LISTEN      -               
udp        0      0 127.0.0.1:11211         0.0.0.0:*                           -               
```

**Comment by ansibot** [comment]
From @geerlingguy on 2015-11-12T20:22:44Z

Final follow-up: Disabling TTY for the Docker exec command and fixing the underlying problem (in this case, not using `service: state=restarted` on the first run, for some reason) seems to have fixed the build: https://travis-ci.org/geerlingguy/ansible-role-memcached/builds/159764487 (though no pretty colors without the TTY :(

**Comment by ansibot** [comment]
From @geerlingguy on 2015-11-12T20:22:44Z

Moved this issue to the main repo: https://github.com/ansible/ansible/issues/19763

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (108/100 (108%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext4.iad.github.net ([192.30.252.195]:34248 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drlds-004Dww-4A
	for greg@biomassiv.es; Tue, 12 Sep 2017 08:54:20 -0500
Date: Tue, 12 Sep 2017 06:35:15 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505223684;
	bh=nKf+w+XanzCBIcNHM4J/G6Kx5g6N1+CTqvc+NL86eeI=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=kW8noSoMpStBiZolm50zq+vrXtROG8RWKB0Rz6DIq2HtPH0KE1q20Fjq3FdFIGM2S
	 eZUje26yK2d4rj300Z63lGy/0FROEVWs+ugD2tlUVDr2iqxVMxiboq+CPXyR1IF8oC
	 +o0clRaML2WLfqTyBE28DRl+sqUpdYt8TbfjzDYs=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (107/100 (107%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext5.iad.github.net ([192.30.252.196]:60105 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drldf-004Ds5-Bn
	for greg@biomassiv.es; Tue, 12 Sep 2017 08:54:07 -0500
Date: Tue, 12 Sep 2017 06:35:18 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505223579;
	bh=lwKo+G4Lrp3cMfflwnVyDkwWCF4kv3Xndy56MuPD06E=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=QJQ23uXDG/ceYsYLHB1JJUeqtO7pwvm8NbUDHNNj2fK0IJv6oKPF0SZUXB74XotwN
	 2mtMGlFpZsuIz+8nHMEH+mvmqVp5hCnaoTJAe54tDXi6NpYlTbsqM8JpjE7iPwogyA
	 9isFgX2zBv1GJA8K2ItUcX1U6fN+XyCp8GaWrRHQ=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (114/100 (113%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 6823 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext2.iad.github.net ([192.30.252.193]:45541 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drliw-004GW9-6z
	for greg@biomassiv.es; Tue, 12 Sep 2017 08:59:34 -0500
Date: Tue, 12 Sep 2017 06:35:18 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505223554;
	bh=BUF3k3I1y8R51UYiye1DXZsidbaM2KdjuwKJDsAdPJo=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=Sk1oIXou/196utwg8o0lNJnwKImTQtDe4ndftPObKCLkUaWEGxPztK+NmCyulmOWm
	 SbuxJrvE8mqPOIDXuB5OB6TYSq9iYjdUjFXjaDu3mbwjc3ggREf+vD9chlYiQcfAiX
	 +cqC0w95nTfpA/ikJv8fnwGP7DAZVFB636ECIu4o=

**Comment by ansibot** [comment]
@ansibot Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by infrascripting** [comment]
Related: https://github.com/ansible/ansible/issues/18305#issuecomment-327488669

**Comment by jctanner** [comment]
Duplicate of https://github.com/ansible/ansible/issues/19763
